aries/aries_vcx_core/src/wallet2/mod.rs

```diff
@@ -0,0 +1,77 @@
+use async_trait::async_trait;
+use futures::stream::BoxStream;
+
+use crate::errors::error::VcxCoreResult;
+
+pub enum SigType {
+    EdDSA,
+    ES256,
+    ES256K,
+    ES384,
+}
+
+impl From<SigType> for &str {
+    fn from(value: SigType) -> Self {
+        match value {
+            SigType::EdDSA => "eddsa",
+            SigType::ES256 => "es256",
+            SigType::ES256K => "es256k",
+            SigType::ES384 => "es384",
+        }
+    }
+}
+
+#[async_trait]
+pub trait Wallet: RecordWallet + DidWallet {}
+
+#[async_trait]
+pub trait DidWallet {
+    type DidAttrs;
+    type CreatedDid;
+    type DidKey;
+    type KeyAttrs;
+
+    async fn create_key(&self, key_attrs: Self::KeyAttrs) -> VcxCoreResult<()>;
+
+    async fn create_did(&self, attrs: Self::DidAttrs) -> VcxCoreResult<Self::CreatedDid>;
+
+    async fn did_key(&self, did: &str) -> VcxCoreResult<Self::DidKey>;
+
+    async fn replace_did_key(&self, did: &str) -> VcxCoreResult<Self::DidKey>;
+
+    async fn sign(
+        &self,
+        verkey_name: &str,
+        msg: &[u8],
+        sig_type: SigType,
+    ) -> VcxCoreResult<Vec<u8>>;
+
+    async fn verify(
+        &self,
+        vk: &str,
+        msg: &[u8],
+        signature: &[u8],
+        sig_type: SigType,
+    ) -> VcxCoreResult<bool>;
+}
+
+#[async_trait]
+pub trait RecordWallet {
+    type Record;
+    type RecordId;
+    type FoundRecord;
+    type SearchFilter;
+
+    async fn add_record(&self, record: Self::Record) -> VcxCoreResult<()>;
+
+    async fn get_record(&self, id: &Self::RecordId) -> VcxCoreResult<Self::FoundRecord>;
+
+    async fn update_record(&self, update: Self::Record) -> VcxCoreResult<()>;
+
+    async fn delete_record(&self, id: &Self::RecordId) -> VcxCoreResult<()>;
+
+    async fn search_record(
+        &self,
+        filter: Self::SearchFilter,
+    ) -> VcxCoreResult<BoxStream<VcxCoreResult<Self::FoundRecord>>>;
```

Comment:
I am confused about the return type here.

1. The implementation's `SearchFilter` contains `limit`, option, but this function seems to return a single record? Can't multiple records match the filter and if so, what is the behavior if they do?
2. What is the reasoning behind using `BoxStream` (and thus hardcoding `futures` into the signature)?
3. What is the reasoning behind the `Result` "nesting"?

Reply:
1. This method returns a stream in `Result`, streams can contain multiple items. Multiple records can match the filter, I would expect all matches to be included in result stream.
2. This method signature was "[borrowed](https://github.com/hyperledger/aries-vcx/pull/1042/files#diff-9e6a1b11d50e8360d9670a527baa801351e1942b35e7551579c61500a1408749R88)" so I think the idea was to have lazily evaluated search.
3. When fetching each record lazily, each fetch can fail.

The "eager" variant could simplify the method signature into something like `-> VcxCoreResult<Vec<Self::FoundRecord>>`

Reply:
It now simplified.